### PR TITLE
DropWizard Integration

### LINKED ISSUE #21: Make Tree Drop from Explorer recursive
- [x] If a directory is dropped open a Wizard that asks if the user wants to copy all nested files/dirs.
- [x] Also add choice of shallow (first level children) or full recursive

### PR DESCRIPTION
**Pull request by valentinpurrucker**
Closes #21 

- Adds a wizard that lets the user choose betwenn different copying strategies when dropping directories in the explorer.

### CODE REVIEW

**Comment by timbittner** [comment]
Seems to be working and behaving as expected. I tested:
- Dropping Multiple Folders and Files at once with each copy option.
- Dropping a single Folder
- Dropping a single Empty Folder
- Dropping Multiple Empty Folders
- Dropping a single File
- Dropping a few files
- Dropping a Folder containing only a single empty folder

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java [hunk 1]
```diff
@@ -323,6 +340,19 @@ public void toggleView(@UIEventTopic(E4CEventTable.EVENT_VIEW_TOGGLE) Object o)
 		services.eventBroker.send(E4CEventTable.EVENT_REFRESH_PROJECT_VIEWER, null);
 	}
 
+	/**
+	 * Subscribing on the drop element event
+	 */
+	@Inject
+	@Optional
+	public void dropElements(@UIEventTopic(E4CEventTable.EVENT_DROP_ELEMENT_IN_EXPLORER) Object o) {
+		System.out.println("About to copy a directory");
```
Comment (timbittner) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java]: Remove Debug print statement :)

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/wizards/ImportDirectoryPage.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * Represents a Wizard Page which lets the user choose to copy the content of a
+ * directory or only the directory itself (empty).
+ */
+public class ImportDirectoryPage extends WizardPage {
+
+	private IEclipseContext context;
+
+	private CopyStrategy copyStrategy = CopyStrategy.EMPTY;
+
+	public ImportDirectoryPage(String pageName, Path path, IEclipseContext context, ImageDescriptor imgDescriptor) {
```
Comment (timbittner) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/wizards/ImportDirectoryPage.java:42]: Javadoc

**Review by timbittner — COMMENTED** [review]
Minimal Changes, DocString, Debug print.
While you're at it: try using eclipse formatting (STRG+SHIFT+F) on java files for nice and carefree formatting. (Except if that breaks your git and all files are then overwritten 😆  )

File: de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
+	
+	public static final String EVENT_DROP_ELEMENT_IN_EXPLORER = "DROP_ELEMENT_IN_EXPLORER";
+	
+	public static final String UPDATE_STATUS_BAR = "UPDATE_STATUS_BAR";
+
+	public static final String EVENT_RENAME_PROJECT_EXPLORER_ITEM = "EVENT_RENAME_PROJECT_EXPLORER_ITEM";
+	
+	public static final String EVENT_FILTER_CHANGED = "EVENT_FILTER_CHANGED";
+	public static final String EVENT_VIEW_TOGGLE = "EVENT_VIEW_TOGGLE";
+}
```
Comment (Jooshua) [inline de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java:19]: This git bug is really annoying -.-

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java [hunk 4]
```diff
@@ -53,6 +55,7 @@ public CustomTreeCell(WorkspaceFileSystem workspaceFileSystem, FileImageProvider
 
 			final Dragboard db = event.getDragboard();
 			boolean success = false;
+			List<java.io.File> directories = new ArrayList<File>();
```
Comment (Jooshua) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java]: We can do just `File`?

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java [hunk 5]
```diff
@@ -1,25 +1,27 @@
...
-}
+package de.tu_bs.cs.isf.e4cf.parts.project_explorer.stringtable;
+
+import java.nio.file.Paths;
+
+public class FileTable {
+	public static final String PROJECT_EXPLORER_VIEW_FXML = Paths.get("ui", "view", "ProjectExplorerView.fxml")
+			.toString();
+
+	public static final String COPY_DIRECTORY_PAGE_FXML = Paths.get("ui", "view", "CopyDirectoryView.fxml").toString();
```
Comment (Jooshua) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java]: See comment in the #37. Also wondering about the line breaking here

**Review by kyshidar — APPROVED** [review]

**Review by Jooshua — CHANGES_REQUESTED** [review]
Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...

File: de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java (hunk 3 above)
Comment (timbittner) [reply de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java:19]: But it seems to happen less and less :D

Comment (valentinpurrucker) [reply de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java:19]: Yeah i'm sorry :D :D

**Comment by timbittner** [comment]
> Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...

You are talking about in-tree (ProjectExplorer) drag and drop right?
Dragging from Windows File Explorer should definetly always COPY.

I like @Jooshua's Idea of having a Little Checkbox titled "Move instead of Copy Files" or something like that. Depending on that check the wizards behavior changes. All other in-Tree DnD behavior is still working as expected (move).

That new checkbox needs to be disabled and set to copy mode when invoking the wizard from file system.  That is something that can be derived from the DragSource and I believe we already did something llike that in the TreeCellFactory.

@valipu Do you think this is an easy fix to implement or more challenging and should be its own issue instead of a PR change?

**Comment by valentinpurrucker** [comment]
> > Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...
> 
> You are talking about in-tree (ProjectExplorer) drag and drop right?
> Dragging from Windows File Explorer should definetly always COPY.
> 
> I like @Jooshua's Idea of having a Little Checkbox titled "Move instead of Copy Files" or something like that. Depending on that check the wizards behavior changes. All other in-Tree DnD behavior is still working as expected (move).
> 
> That new checkbox needs to be disabled and set to copy mode when invoking the wizard from file system. That is something that can be derived from the DragSource and I believe we already did something llike that in the TreeCellFactory.
> 
> @valipu Do you think this is an easy fix to implement or more challenging and should be its own issue instead of a PR change?

isn't it easier to just include the mode (copy or move) directly to the `DropElement` @timbittner  ? I think no user expects 

> > Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...
> 
> You are talking about in-tree (ProjectExplorer) drag and drop right?
> Dragging from Windows File Explorer should definetly always COPY.
> 
> I like @Jooshua's Idea of having a Little Checkbox titled "Move instead of Copy Files" or something like that. Depending on that check the wizards behavior changes. All other in-Tree DnD behavior is still working as expected (move).
> 
> That new checkbox needs to be disabled and set to copy mode when invoking the wizard from file system. That is something that can be derived from the DragSource and I believe we already did something llike that in the TreeCellFactory.
> 
> @valipu Do you think this is an easy fix to implement or more challenging and should be its own issue instead of a PR change?

@timbittner i think we can includ it since it is not really much to change

**Comment by valentinpurrucker** [comment]
When i tested dropping a child directory with content to another directroy inside the tree i get a PermissionDeniedException but i don't think thats because of the implementation because i used to get it when we were dropping files in the tree too.

Just to be clear: the directory is still moved / copied correclty even when there is this AccessDeniedExc

**Comment by timbittner** [comment]
Unfortunately while testing your latest comment I found another critical issue:

If I drag a folder onto its parent folder and select the 'move instead of copy' property the folder will be deleted leading to data loss.

**Comment by Jooshua** [comment]
> 
> 
> Unfortunately while testing your latest comment I found another critical issue:
> 
> If I drag a folder onto its parent folder and select the 'move instead of copy' property the folder will be deleted leading to data loss.

Säd :/ I will rereview it after this there is another commit regarding this issue. Also you can now merge the development into this branch.

**Comment by valentinpurrucker** [comment]
Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua

**Comment by timbittner** [comment]
> Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua

Sure 😆 
The original issue seems fixed but when I enable move while selecting 'shallow' or 'folder only' mode I still loose a lot of files.
That leads me to believe that we should enable move-mode only when performing 'deep'-move

**Comment by timbittner** [comment]
Minor Detail: When trying to drop a file into a folder where another file with that name already exists an error dialog pops up, which is fine, but it does not contain any hints on what went wrong 😄

**Comment by valentinpurrucker** [comment]
> > Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua
> 
> Sure 
> The original issue seems fixed but when I enable move while selecting 'shallow' or 'folder only' mode I still loose a lot of files.
> That leads me to believe that we should enable move-mode only when performing 'deep'-move

What do you mean by loosing files in shallow or create only folder? Does that happen when you have multiple files at once or just in general? Just curious 😜

**Comment by Jooshua** [comment]
> 
> 
> > > Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua
> > 
> > 
> > Sure
> > The original issue seems fixed but when I enable move while selecting 'shallow' or 'folder only' mode I still loose a lot of files.
> > That leads me to believe that we should enable move-mode only when performing 'deep'-move
> 
> What do you mean by loosing files in shallow or create only folder? Does that happen when you have multiple files at once or just in general? Just curious 😜

How about some discord voice then? We are discussing this ticket atm...

**Review by Jooshua — APPROVED** [review]

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java [hunk 6]
```diff
@@ -53,6 +57,8 @@ public CustomTreeCell(WorkspaceFileSystem workspaceFileSystem, FileImageProvider
 
 			final Dragboard db = event.getDragboard();
 			boolean success = false;
+			DropMode dropMode = DropMode.COPY;
+			List<File> directories = new ArrayList<File>();
```
Comment (timbittner) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java:69]: This file needs more comments around fields and if-statements to make it easier to understand whats going on.
The code in and of itself is fine though.

**Review by timbittner — APPROVED** [review]

**Pull request merged**